### DISCUSSION TITLE
Generate Stateprovider using riverpod_generator

### DISCUSSION THREAD

**Discussion by wurikiji** [rrousselGit/riverpod#1725] [Q&A] [answered]
I'd like to generate `StateProvider` using riverpod_generator. 
But I found that there are only four types of provider, `provider`, `futureProvider`, `notifier`, and `asyncNotifier`, in the generator code. 
Do I have to use `StateNotifier` or `ChangeNotifier` even for simple state management?

**Comment by TimWhiting** [comment, marked as answer]
StateProvider is simple without the generator. I'm curious why you are wanting to generate it specifically. The generator makes StateNotifier providers simpler by letting you define methods directly on a class and not splitting up the definition of the class and the definition of the provider.

You can freely use regular riverpod and generated riverpod together.

**Reply by rrousselGit** [reply]
In the long run, assuming metaprogramming isn't canceled, the old syntax will likely be deprecated.  
It's far in the future though.

**Comment by rrousselGit** [comment]
There is no plan for now to have a StateProvider equivalent in the code generator.

Making a class and generating a `Notifier` is the recommended solution for now.

**Reply by ened** [reply]
I suggest documenting this on the [website](https://docs-v2.riverpod.dev/docs/about_code_generation). :) 

Just now I ran into:
> Generated providers should only depend on other generated providers. Failing to do so may break rules such as "provider_dependencies".dart(avoid_manual_providers_as_generated_provider_dependency)

Reason being a state provider has been watched.

**Reply by ened** [reply]
@rrousselGit wdyt, is this worth documenting?

**Comment by ValentinVignal** [comment]
There is a package to generated the equivalent of the `StateProvider`: 

https://pub.dev/packages/riverpod_state_provider_annotation
https://pub.dev/packages/riverpod_state_provider_generator

```dart
# main.dart
import 'package:riverpod_state_provider_annotation/riverpod_state_provider_annotation.dart';

part 'main.g.dart';
part 'main.rsp.dart';

@RiverpodStateProvider()
int myState() {
  return 0;
}

ref.watch(myStateProvider);
ref.read(myStateProvider.notifier).state++;
ref.read(myStateProvider.notifier).update((state) => state + 1);

ProviderScope(
  overrides: [
    myStateProvider.overrideWithValue(10),
  ],
)
```

It still has some limitations but for simple usages (usually the case for `StateProvider`), it does the job
